### DISCUSSION TITLE
Wikiquery as a tool to fix issues

### DISCUSSION THREAD

**Discussion by tholzheim** [SmartDataAnalytics/OpenResearch#121] [Ideas]
For example: #119
could be easily analyzed with 
```
wikiquery -s or -q "[[Ordinal::+]]"
```
but this would only give us the page names at this time
Result: 
```
$wikiquery -s or -q "[[Ordinal::+]]" | wc -l
358
```
If we fix https://github.com/WolfgangFahl/py-3rdparty-mediawiki/issues/56 it should be easy to  fix #119.
By applying unix tools such as grep.

**Comment by WolfgangFahl** [comment]
That is an excellent proposal. I also think that we might be quicker in assessing the relevance of properties this way. Could you check what a query would looke like that finds out how often the Event and Event Series relevant properties have been used in OPENRESEARCH? Applying that ask query and then using wikiquery might give us a speed up in progress. Also it would be a "surrogate" for our lamdba approach we could simply apply a unix pipeline of commands and this would not even be limited to python ...

**Comment by tholzheim** [comment]
Which formats should be implemented first?

**Comment by WolfgangFahl** [comment]
csv and json IMHO would best fit with unix tool pipelines. 
```
cut -f 
```
will easily give you the field of a csv or you could use awk
for json there is the tool jq

**Reply by tholzheim** [reply]
I don't know these three but I will try them out.

**Comment by WolfgangFahl** [comment]
I propose you fix the ticket and then i'll show you examples of all three commands. You'll also find some of the usage of these unix tools in the script that gathers samples for the proceedings title parser. see https://github.com/WolfgangFahl/ProceedingsTitleParser/blob/master/scripts/getsamples

**Reply by tholzheim** [reply]
I found this use case of cut and jq within a one line pipeline
```
  cat $sampledir/crossref-*.json | jq .message.items[].title | cut -f2 -d'[' | cut -f2 -d'"' | grep -v "]" | tr -s '\n' > $sampledir/proceedings-crossref.txt
```
https://github.com/WolfgangFahl/ProceedingsTitleParser/blob/b48832e9032e41785f61338f6ff2f5cac91aba0e/scripts/getsamples#L243

Looks like it could be applied straight forward to the results of wikiquery.

**Comment by tholzheim** [comment]
General question: Can invalid property values be queried?
For test purposes I intentionally changed the Ordinal of https://www.openresearch.org/wiki/VLDB_2016 to an invalid value.
An error message is shown in the infobox as hint, that the value is invalid but I can not query the value with wikiquery nor is it shown in the Event series https://www.openresearch.org/wiki/VLDB

**Reply by WolfgangFahl** [reply]
I am not aware of a native SMW solution for this but I think we might be able to simply use wikibackup and unix tools like awk grep and sed to create a toolchain. Effectively we'd end up with some kind of wikilambda that might use python but also other unix tools. To show this off would be good next step. Feel free to chat online with me on this next time you are available.

**Reply by tholzheim** [reply]
The property [Property:Has improper value for](https://www.openresearch.org/wiki/Property:Has_improper_value_for) stores all invalid values. This means we can query for all events that have an invalid ordinal:
```
{{#ask:[[Has improper value for::Ordinal]]|format=ul}}
```
See https://www.openresearch.org/wiki/Issue:Incorrect_Ordinal

**Reply by WolfgangFahl** [reply]
This would also be a good comment for #7 and #71

**Comment by WolfgangFahl** [comment]
Please try this out together with Musaab on the #119 task. Make sure you try out things in your OPENRESEARCH copies! We'll then later try the same approach e.g. for importing from other sources like dblp.

First step:
```
 grep "{{Event" *.wiki -A20 | grep "|Ordinal" | cut -f 2 -d "=" | sort -u 
```

These are the values we have:
```
1
10
10th
11
11th
12
12th
13
13th
14
15
16
16th
17
18
19
2
20
21
22
23
24
25
26
27
28
29
3
30
31
32
33
34
35
36
38
39
4
40
41
41st
44
45
46
5
52
6
68
7
7th
8
9
9th
```

now we need a "not" condition - find the entries that are not conforming to the regexp for numbers. So instead of using the unix tools above you may simply use python to find the problematic ones. Then do a loop over the .wiki files and fix the Ordinal with the lookup in the dictionary mentioned in the ticket. Then do a wikirestore:

So the way to do this is:

1. Wikibackup or wikiquery to get relevant pages
2. Unix-Pipeline to find and fix
3. Wikirestore to apply fix

**Reply by tholzheim** [reply]
Regarding step 2: I have implemented a [python script](https://github.com/SmartDataAnalytics/OpenResearch/files/6030388/ordinal_to_cardinal.zip) to detect and change the wrong ordinal numbers.
Similar to https://github.com/WolfgangFahl/ProceedingsTitleParser/blob/7e52b4e3eae09269464669fe387425b9f6392952/ptp/titleparser.py#L488 it takes a dictionary as input for the lookup as mentioned in #119.
The script is executed as follows:
`$ ./ordinal_to_cardinal.py -d dictionary.yaml -p orth`
> Add `-f` to actually apply the changes.

If the value lookup fails an error message is printed out and the page contend is not changed.
`orth/CQR 2008.wiki:      Lookup failed! CQR is missing in the dictionary. `
What is the desired behavior for invalid ordinal values that can't be resolved?
Like this one https://www.openresearch.org/wiki/CQR_2008

###  Limitations
Currently  the script searches only for the first occurrence of the property Ordinal.


## Example execution
```
$ ./ordinal_to_cardinal.py -d dictionary.yaml -p orth
orth/ICACM 2021.wiki:    4th will changed to 4.
orth/ICNST 2021.wiki:    4th will changed to 4.
orth/ICIM 2021.wiki:     7th will changed to 7.
orth/2021 ICIMP.wiki:    4th will changed to 4.
orth/IEEA 2021.wiki:     10th will changed to 10.
orth/BUSTECH 2021.wiki:  11th will changed to 11.
orth/ICBBT 2021.wiki:    13th will changed to 13.
orth/ICCRE 2021.wiki:    6th will changed to 6.
orth/ICEDA 2021.wiki:    1st will changed to 1.
orth/PATTERNS 2021.wiki:         13th will changed to 13.
orth/AITC 2021.wiki:     3rd will changed to 3.
orth/ALENEX 2019.wiki:   21st will changed to 21.
orth/BAAL 2008.wiki:     41st will changed to 41.
orth/SCAP 2021.wiki:     6th will changed to 6.
orth/ICMAE 2021.wiki:    12th will changed to 12.
orth/ACSS 2021.wiki:     3rd will changed to 3.
orth/ICGET 2021.wiki:    6th will changed to 6.
orth/ICCAI 2021.wiki:    7th will changed to 7.
orth/ICENT 2021.wiki:    3rd will changed to 3.
orth/ICSSE 2021.wiki:    4th will changed to 4.
orth/ADAPTIVE 2019.wiki:         11th will changed to 11.
orth/SPML 2021.wiki:     4th will changed to 4.
orth/AHMIS 2021.wiki:    2nd will changed to 2.
orth/WCSE 2021.wiki:     11th will changed to 11.
orth/CLOUD COMPUTING 2021.wiki:  12th will changed to 12.
orth/ICEAT 2021.wiki:    1st will changed to 1.
orth/ICMHI 2021.wiki:    5th will changed to 5.
orth/CGEEE 2021.wiki:    4th will changed to 4.
orth/ICMEE 2021.wiki:    6th will changed to 6.
orth/ADAPTIVE 2017.wiki:         9th will changed to 9.
orth/CVIT 2021.wiki:     2nd will changed to 2.
orth/CONTENT 2021.wiki:  13th will changed to 13.
orth/ICECT 2021.wiki:    3rd will changed to 3.
orth/INFOCOMP 2021.wiki:         11th will changed to 11.
orth/ICAFM 2021.wiki:    6th will changed to 6.
orth/CCAI 2021.wiki:     1st will changed to 1.
orth/WSSE 2021.wiki:     3rd will changed to 3.
orth/ICBEM 2021.wiki:    11th will changed to 11.
orth/MSIE 2021.wiki:     3rd will changed to 3.
orth/ICBIP 2021.wiki:    6th will changed to 6.
orth/ICBCB 2021.wiki:    9th will changed to 9.
orth/AACL 2008.wiki:     8th will changed to 8.
orth/ICMSSP 2021.wiki:   6th will changed to 6.
orth/ICCEMS 2021.wiki:   6th will changed to 6.
orth/COGNITIVE 2021.wiki:        13th will changed to 13.
orth/ICMESM 2021.wiki:   6th will changed to 6.
orth/SPIES 2021.wiki:    3rd will changed to 3.
orth/ICIEA 2021.wiki:    8th will changed to 8.
orth/ICAER 2021.wiki:    7th will changed to 7.
orth/IECC 2021.wiki:     3rd will changed to 3.
orth/ICIMP 2021.wiki:    16th will changed to 16.
orth/FUTURE COMPUTING 2021.wiki:         13th will changed to 13.
orth/WSCE 2021.wiki:     4th will changed to 4.
orth/ADAPTIVE 2020.wiki:         12th will changed to 12.
orth/ICDMA 2021.wiki:    7th will changed to 7.
orth/ACMME 2021.wiki:    9th will changed to 9.
orth/ICEEA 2021.wiki:    12th will changed to 12.
orth/ICBDA 2021.wiki:    6th will changed to 6.
orth/ICGDA 2021.wiki:    4th will changed to 4.
orth/BDSIC 2021.wiki:    3rd will changed to 3.
orth/ICRES 2021.wiki:    3rd will changed to 3.
orth/IEAI 2021.wiki:     2nd will changed to 2.
orth/ICIAI 2021.wiki:    5th will changed to 5.
orth/ICBDC 2021.wiki:    6th will changed to 6.
orth/BIOTC 2021.wiki:    3rd will changed to 3.
orth/ICoMS 2021.wiki:    4th will changed to 4.
orth/ASIP 2021.wiki:     3rd will changed to 3.
orth/ICFIE 2021.wiki:    8th will changed to 8.
orth/3IA 2009.wiki:      12th will changed to 12.
orth/IVSP 2021.wiki:     3rd will changed to 3.
orth/ICISPC 2021.wiki:   5th will changed to 5.
orth/AICCC 2021.wiki:    4th will changed to 4.
orth/CQR 2008.wiki:      Lookup failed! CQR is missing in the dictionary. 
orth/SERVICE COMPUTATION 2021.wiki:      13th will changed to 13.
orth/ICSED 2021.wiki:    3rd will changed to 3.
orth/ICoIAS'2021.wiki:   4th will changed to 4.
orth/ICCMP 2021.wiki:    7th will changed to 7.
orth/AEEES 2021.wiki:    3rd will changed to 3.
orth/ADIP 2021.wiki:     3rd will changed to 3.
orth/BDE 2021.wiki:      3rd will changed to 3.
orth/ADAPTIVE 2018.wiki:         10th will changed to 10.
orth/ICPAM 2021.wiki:    10th will changed to 10.
orth/COMPUTATION TOOLS 2021.wiki:        12th will changed to 12.

```




Regarding step 3: Does there already exist a wikirestore tool or is this related to https://github.com/WolfgangFahl/py-3rdparty-mediawiki/issues/44

**Comment by WolfgangFahl** [comment]
Good work! That is the right direction. Indeed we neeed the wikirestore fix as a prerequisite. Also it would be good to have the wikiquery input "piped" in to the solution. We have to think about a standard way to do this e.g. stdin/stdout using filenames or json info or similar means. When we are done with this step the next step is trying to extract the ordinal from the title as the proceedings title parser does. We could again take a snippet from the source code or give the ptp an API for this.

**Reply by MusaabKh** [reply]
I was working on extracting the ordinal from the title aside from fixing the ordinal and wanted to make sure of the title will have the number always in the ordinal form or is the cardinal form also expected in the title?

**Reply by tholzheim** [reply]
Wikirestore should work now. You find a small documentation here https://confident.dbis.rwth-aachen.de/orth/index.php?title=Wikirestore

**Comment by MusaabKh** [comment]
wikirestore now has a new functionality of reading input from STD IN. Pipeline to fix ordinal numbers issue is as follows:
1) wikibackup to get files into the system
2) grep files where ordinal field exists.
Pipe such files to the following python file: 
[Python_Ordinal.zip](https://github.com/SmartDataAnalytics/OpenResearch/files/6065394/Python_Ordinal.zip)
4) pipe the output from the python file to wikirestore.
Running command:
`grep Ordinal C:/Users/92031/wikibackup/ormk -l -r | python ordinal_to_cardinal.py -stdin -d '../dictionary.yaml' -ro -f | wikirestore -t ormk -stdinp -ui`

grep can later be replaced by wikiquery if it returns the page names in STD in.

Wikirestore can also take a listFile argument now and restore all the filenames from the given file.

**Comment by WolfgangFahl** [comment]
Please keep in mind we need to check the types being used in OpenResearch and fix https://github.com/WolfgangFahl/py-3rdparty-mediawiki/issues/5 accordingly. We do not need all types but the relevant ones e.g.  bool and geo coordinates might bee needed see also https://github.com/BITPlan/com.bitplan.simplegraph/blob/master/simplegraph-smw/src/main/java/com/bitplan/simplegraph/smw/SmwSystem.java
